language/stdlib/modules/GenesisState.move

```diff
@@ -0,0 +1,145 @@
+address 0x1 {
+
+/// This file implements a state variable to track the status of genesis.
+/// It is in a separate leave module, because it will be used by lots of
+/// other modules to check whether functions are being called during the
+/// genesis process or not. Putting it in Genesis would introduce a ton
+/// of dependencies, but using GenesisState will not.
+module GenesisState {
+    use 0x1::Signer;
+    use 0x1::CoreAddresses;
+
+    /// A singleton resource that stores the state of the genesis process
+    /// on genesis_address()
+    resource struct GenesisState {
```

Comment:
My initial thought was that this `genesis_complete` field could be able be elided, and that this might make the other functions simpler: 
* `begin_genesis` publishes the resource, 
* `is_during_genesis` checks for the existence of the resource,
* `end_genesis` unpublishes the resource, and 
* `is_after_genesis` checks that the resource doesn't exist. 

The only problem with this is that it looks like doing this might make the `TransBeginToAfter` impossible to  prove? Since  if the resource is unpublished nothing prevents `begin_genesis` from being called again. 

If that's the case, this looks good to me. Only thing might be getting rid of `genesis_address` and moving that in to the `CoreAddresses` module.

Reply:
"The only problem" was the problem I was trying to solve.  I really didn't want people to be able to re-enter genesis.

I agree about CoreAddresses.

Reply:
Ah ok awesome. 

One other thought I had with this, is that this boolean separates it into three phases (before, during, after). Would it be ok to instead only have two? e.g. (genesis, after genesis)? Then `is_genesis` looks for the non-existence, and `is_after_genesis` looks for the existence. This would have the irreversible property that you want, but there is no distinction between "before" and "during" genesis.